### PR TITLE
Adding support for HTTP 

### PR DESCRIPTION
**Pull request by rtpro**

### CODE REVIEW

File: application/constants/templates.py [hunk 1]
```diff
@@ -2,6 +2,10 @@
 Templates related constants.
 """
 from .base_enum import StrEnum
+from typing import Union
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:5]: Please sort import in right order. There is convention of sorting imports:

1. future imports `from __future__ import some_feature_that_supported_in_higher_python_version`(we don't have such here).
2. imports from standard library.
3. Third party imports Like: `from pydantic import Field`.
4. project imports. Like: `from service_hub import some_thing`(we don't have such example here)
5. local import like: `from .base_enum import StrEnum`.

Additional to it. In this project we have convention that one line one import.

Comment (vadim-zabolotniy) [reply application/constants/templates.py:5]: You can see import example here: https://github.com/JovianX/Service-Hub/blob/main/application/api/v1/schemas/applications.py

File: application/constants/templates.py [hunk 2]
```diff
@@ -2,6 +2,10 @@
...
+from pydantic import Field, BaseModel, constr
+from .base_enum import StrEnum
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:7]: Duplicating import. Have you setup linting in your IDE?

File: application/constants/templates.py [hunk 3]
```diff
@@ -43,3 +49,51 @@ class HookTypes(StrEnum):
     Template application hook types.
     """
     kubernetes_job = 'kubernetes_job'
+
+
+class http(StrEnum):
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:54]: PEP8. Class names should be in  Pascal case. It like Camel case but first later is capital.

File: application/constants/templates.py [hunk 4]
```diff
@@ -43,3 +49,51 @@ class HookTypes(StrEnum):
...
+    """
+    Template HTTP component types.
+    """
+    url = 'url'
+    method = 'method'
+    headers = 'headers'
+    parameters = 'parameters'
+
+class httpMethods(StrEnum):
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:63]: PEP8. Pascal case.

File: application/constants/templates.py [hunk 5]
```diff
@@ -43,3 +49,51 @@ class HookTypes(StrEnum):
...
+    Template HTTP component methods.
+    """
+    get = 'GET'
+    post = 'POST'
+    put = 'PUT'
+    patch = 'PATCH'
+    delete = 'DELETE'
+
+
+class httpDeploy(BaseModel):
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:74]: PEP8 Pascal case.

File: application/constants/templates.py [hunk 6]
```diff
@@ -43,3 +49,51 @@ class HookTypes(StrEnum):
...
+    """
+    Deploy schema for HTTP type components.
+    """
+    url: constr(min_length=1, strip_whitespace=True) = Field(description='URL for HTTP deployment', example='http://example.com')
+    method: Union[httpMethods, constr(min_length=1, strip_whitespace=True)] = Field(default=httpMethods.get, description='HTTP method for deployment', example=httpMethods.post)
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:79]: If I understood what you tried to achieve you could do just `method: httpMethods`

File: application/constants/templates.py [hunk 7]
```diff
@@ -43,3 +49,51 @@ class HookTypes(StrEnum):
...
+    put = 'PUT'
+    patch = 'PATCH'
+    delete = 'DELETE'
+
+
+class httpDeploy(BaseModel):
+    """
+    Deploy schema for HTTP type components.
+    """
+    url: constr(min_length=1, strip_whitespace=True) = Field(description='URL for HTTP deployment', example='http://example.com')
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:78]: Please use `HttpUrl` for proper url validation. See next example here: https://github.com/JovianX/Service-Hub/blob/c6e80eb9e6a74462ba14ae036252407d66e96982/application/api/v1/schemas/helm.py#L23

File: application/constants/templates.py [hunk 8]
```diff
@@ -43,3 +49,51 @@ class HookTypes(StrEnum):
...
+    method: Union[httpMethods, constr(min_length=1, strip_whitespace=True)] = Field(default=httpMethods.get, description='HTTP method for deployment', example=httpMethods.post)
+    headers: Optional[dict] = Field(description='Headers for HTTP deployment', example={'Content-Type': 'application/json'})
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:80]: We have Python version that allows to use modern more simpler type notation. So ` headers: Optional[dict]` can be replaced by ` headers: dict | None` or I assume we can go further and define it like: ` headers: dict[str, str] | None`.

Comment (vadim-zabolotniy) [reply application/constants/templates.py:80]: Also I think it should be `dict[str, str]`. I mean not optional but empty dict is acceptable value. But this is optional comment.

File: application/constants/templates.py [hunk 9]
```diff
@@ -43,3 +49,51 @@ class HookTypes(StrEnum):
...
+    parameters: Optional[dict | list[dict]] = Field(description='Parameters for HTTP deployment', example=[{'param1': 'value1'}])
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:81]: Modern type notation. `Optional[dict | list[dict]]` can be replaced by `dict | list[dict] | None`. 

But! I think it can be replaced by `dict[str, str]`.

1. Same reason as for `headers`. What is the point of supporting 2 values for representing same meaning(`None` and `{}`)?
2. What is the point of supporting format `[{'param1': 'value1'}]`? How you will handle case when user pass:
```
[
    {'param1': 'value1'},
    {'param1': 'value2'}
]
```
?

File: application/constants/templates.py [hunk 10]
```diff
@@ -43,3 +49,51 @@ class HookTypes(StrEnum):
...
+
+class httpStatus(BaseModel):
+    """
+    Status schema for HTTP type components.
+    """
+    url: constr(min_length=1, strip_whitespace=True) = Field(description='URL for HTTP status', example='http://example.com/status')
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:87]: Again comment regarding build in url support.

File: application/constants/templates.py [hunk 11]
```diff
@@ -43,3 +49,51 @@ class HookTypes(StrEnum):
...
+class httpDeploy(BaseModel):
+    """
+    Deploy schema for HTTP type components.
+    """
+    url: constr(min_length=1, strip_whitespace=True) = Field(description='URL for HTTP deployment', example='http://example.com')
+    method: Union[httpMethods, constr(min_length=1, strip_whitespace=True)] = Field(default=httpMethods.get, description='HTTP method for deployment', example=httpMethods.post)
+    headers: Optional[dict] = Field(description='Headers for HTTP deployment', example={'Content-Type': 'application/json'})
+    parameters: Optional[dict | list[dict]] = Field(description='Parameters for HTTP deployment', example=[{'param1': 'value1'}])
+
+class httpStatus(BaseModel):
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:83]: I think I understood what you tried to achieve. You tried to create validation schema for each request. The only distinguish is `default=httpMethods.<method>` for each case. Please investigate next example https://github.com/JovianX/Service-Hub/blob/main/application/services/kubernetes/schemas.py Main trick does:
```
Entities = Annotated[ClusterRoleEntity |
                     <...>
                     ServiceAccountEntity |
                     StatefulSetEntity,
                     Field(discriminator='kind')]
```

File: application/crud/events.py [hunk 12]
```diff
@@ -24,7 +24,7 @@ async def list_for_application(self, organization_id: int, application_id: int)
             self.table.organization_id == organization_id,
             self.table.category == EventCategory.application,
             self.table.data['application_id'].as_integer() == application_id
-        )
+        ).order_by(self.table.created_at.desc())
```
Comment (vadim-zabolotniy) [inline application/crud/events.py:27]: Nice catch!

File: application/managers/applications.py [hunk 13]
```diff
@@ -269,18 +274,38 @@ async def get_component_health(self, component: Component, application: Applicat
...
-                application.context_name,
-                application.namespace,
-                component.name
-            )
-            health_status = component_health_status['status']
-            details = component_health_status['details']
-        except ReleaseNotFoundException:
-            health_status = ReleaseHealthStatuses.unhealthy
-            details = {}
+        match component.type:
```
Comment (vadim-zabolotniy) [inline application/managers/applications.py:277]: Nice one! I like it!
Consider to add `case _:` to get rid of unexpected behavior.

File: application/managers/applications.py [hunk 14]
```diff
@@ -269,18 +274,38 @@ async def get_component_health(self, component: Component, application: Applicat
...
+                        component_name=component.name,
+                        url=component.status.url,
+                        method=component.status.method,
+                        headers=component.status.headers,
+                        parameters=component.status.parameters,
+                        dry_run=False
+                    )
+                    health_status = ReleaseHealthStatuses.healthy
+                    details = component_health_status
+                except Exception as error:
```
Comment (vadim-zabolotniy) [inline application/managers/applications.py:306]: Please never try to catch all exceptions. Always catch only what you want to catch. Or at least propagate exception further by calling `raise` after 308 line so other part of project can be averred about occurred exceptions.

File: application/managers/applications.py [hunk 15]
```diff
@@ -336,28 +361,48 @@ async def install_component(self, component: Component, application: Application
...
+                        organization=organization,
+                        context_name=context_name,
+                        component_name=component.name,
+                        url=component.deploy.url,
+                        method=component.deploy.method,
+                        headers=component.deploy.headers,
+                        parameters=component.deploy.parameters,
+                        dry_run=dry_run
+                    )
+                except Exception as error:
```
Comment (vadim-zabolotniy) [inline application/managers/applications.py:400]: Again  you trying to intercept to common exception.

File: application/managers/applications.py [hunk 16]
```diff
@@ -383,21 +428,40 @@ async def uninstall_component(self, application: Application, component: Compone
...
+                        organization=application.organization,
+                        context_name=application.context_name,
+                        component_name=component.name,
+                        url=component.delete.url,
+                        method=component.delete.method,
+                        headers=component.delete.headers,
+                        parameters=component.delete.parameters,
+                        dry_run=dry_run
+                    )
+                except Exception as error:
```
Comment (vadim-zabolotniy) [inline application/managers/applications.py:460]: To common exception catching.

File: application/managers/http/manager.py [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+
+    ############################################################################
+    # Http Requests
+    ############################################################################
+
+
+    async def http_request(
+        self, organization: Organization, context_name: str, component_name: str, url: str, method: str,
+        parameters: dict, headers: dict = None, dry_run: bool = False
```
Comment (vadim-zabolotniy) [inline application/managers/http/manager.py:39]: So `parameters` can be not only `dict`. Reflect it in type notation `parameters: dict | list`

File: application/managers/http/manager.py [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+    ) -> str:
+        """
+        Send an HTTP request to the specified URL.
+        """
+        # Convert to a dict if a list of dicts provided
+        if isinstance(parameters, list):
+            parameters = {k: v for d in parameters for k, v in d.items()}
+
+        # Create an HTTP client
+        async with httpx.AsyncClient() as client:
```
Comment (vadim-zabolotniy) [inline application/managers/http/manager.py:49]: Nice! That is why we choose it!

File: application/managers/http/manager.py [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+            # Prepare the request
+            if method.lower() == 'get':
```
Comment (vadim-zabolotniy) [inline application/managers/http/manager.py:51]: It is working but where is a cute match case?
Also we have [HTTPMethods](https://github.com/JovianX/Service-Hub/blob/c6e80eb9e6a74462ba14ae036252407d66e96982/application/constants/common.py#L10C7-L10C18) enum for http methods. But furthermore you have defined [httpMethods](https://github.com/JovianX/Service-Hub/pull/283/files#diff-3716243d060a9d4e5f05705851171a492c3b8f834f1a7f58ff4611f41a372527R63) and now comparing it to bare stings.

File: application/managers/http/manager.py [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+                response = await client.patch(url, params=parameters, headers=headers)
+            else:
+                raise ValueError(f"Unsupported method: {method}")
+
+        # Check the response status code
+        if not 200 <= response.status_code < 300:
+            raise Exception(f"Request failed with status {response.status_code}")
+
+        # Return the response text
+        # print(response.text)
```
Comment (vadim-zabolotniy) [inline application/managers/http/manager.py:69]: Remove it. 
Please set up our dev environment. Linting formatting and debugging.

File: application/requirements.txt [hunk 21]
```diff
@@ -12,7 +12,7 @@ httpx-oauth
 kubernetes
 kubernetes_asyncio==22.6.5 #  Version locked because of issue https://github.com/kubernetes-client/python/issues/1718 and https://github.com/kubernetes-client/python/issues/1751
 mergedeep
-procrastinate
+procrastinate==0.33.0
```
Comment (vadim-zabolotniy) [inline application/requirements.txt:15]: Looks like you didn't merged my changes with requirements versions locking.

File: application/managers/http/manager.py [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+from constants.events import EventCategory
+from managers.events import EventManager
+from managers.events import get_event_manager
+from managers.kubernetes import K8sManager
+from managers.organizations.manager import OrganizationManager
+from managers.organizations.manager import get_organization_manager
+from models.organization import Organization
+from schemas.events import EventSchema
+from services.helm.facade import HelmService
+import httpx
```
Comment (vadim-zabolotniy) [inline application/managers/http/manager.py:17]: Import order.

File: application/requirements.txt [hunk 23]
```diff
@@ -12,7 +12,7 @@ httpx-oauth
 kubernetes
```
Comment (vadim-zabolotniy) [inline application/requirements.txt:12]: Were is requirement for `httpx`?

File: application/schemas/templates/components.py [hunk 24]
```diff
@@ -5,20 +5,25 @@
...
 class Component(BaseModel):
     """
     Application component schema.
     """
     name: constr(min_length=1, strip_whitespace=True) = Field(description='Helm release name', example='vault')
-    type: ComponentTypes = Field(description='Type of applicatoin component.', example=ComponentTypes.helm_chart)
+    type: ComponentTypes = Field(description='Type of application component.', example=ComponentTypes.helm_chart)
     enabled: bool | TemplateVariable | None = Field(description='Is component must be installed.', default=True)
-    chart: constr(min_length=1, strip_whitespace=True) = Field(description='Helm chart name.', example='bitnami/redis')
+    chart: constr(min_length=1, strip_whitespace=True) | None = Field(description='Helm chart name.', example='bitnami/redis')
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:26]: Aha so you know about modern type notation! -_-

File: application/schemas/templates/components.py [hunk 25]
```diff
@@ -28,6 +33,26 @@ class Component(BaseModel):
                     'a higher priority.',
     )
 
+    deploy: Optional[httpDeploy] = Field(description='Deployment details for HTTP type components')
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:36]: If know about modern type notation why you did this?

File: application/schemas/templates/components.py [hunk 26]
```diff
@@ -28,6 +33,26 @@ class Component(BaseModel):
...
+    status: Optional[httpStatus] = Field(description='Status details for HTTP type components')
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:37]: And this...

File: application/schemas/templates/components.py [hunk 27]
```diff
@@ -28,6 +33,26 @@ class Component(BaseModel):
...
+    delete: Optional[httpDelete] = Field(description='Deletion details for HTTP type components')
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:38]: And this.

File: application/schemas/templates/components.py [hunk 28]
```diff
@@ -5,20 +5,25 @@
...
 from constants.templates import ComponentTypes
+from constants.templates import httpDeploy
+from constants.templates import httpStatus
+from constants.templates import httpDelete
 
 from .validators import TemplateVariable
 
+from typing import Optional
 
 class Component(BaseModel):
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:19]: I think you need to split component definition into two(for now) different models. See this [comment](https://github.com/JovianX/Service-Hub/pull/283/files#r1525485019) for more details how to do it. `discriminator` should by `type`.

Comment (vadim-zabolotniy) [reply application/schemas/templates/components.py:19]: Otherwise it is mess of helm and http components requirements.

File: application/managers/http/manager.py [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+        self.organization_manager = organization_manager
+        self.event_manager = event_manager
+
+
+    ############################################################################
+    # Http Requests
+    ############################################################################
+
+
+    async def http_request(
```
Comment (vadim-zabolotniy) [inline application/managers/http/manager.py:37]: I see that method parameters: ` organization context_name component_name dry_run` are unused. I think you can delete them.

File: application/managers/http/manager.py [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+            elif method.lower() == 'post':
+                response = await client.post(url, params=parameters, headers=headers)
+            elif method.lower() == 'put':
+                response = await client.put(url, params=parameters, headers=headers)
+            elif method.lower() == 'delete':
+                response = await client.delete(url, params=parameters, headers=headers)
+            elif method.lower() == 'patch':
+                response = await client.patch(url, params=parameters, headers=headers)
+            else:
+                raise ValueError(f"Unsupported method: {method}")
```
Comment (vadim-zabolotniy) [inline application/managers/http/manager.py:62]: I think you need to create new exception module with exception related to http manager in `application/exceptions/<...>` and not raise to common `ValueError`.

Also I think you need to intercept exceptions from `httpx` and convert it to http manager's exceptions. Don't shift the internal implementation guts on end user.

File: application/schemas/templates/components.py [hunk 31]
```diff
@@ -28,6 +33,26 @@ class Component(BaseModel):
...
+        deploy = values.get('deploy')
+        delete = values.get('delete')
+        if type_ == ComponentTypes.http:
+            if deploy is None:
+                raise ValueError('deploy is required when type is http')
+            if delete is None:
+                raise ValueError('delete is required when type is http')
+        return values
+
+    def __hash__(self) -> str:
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:53]: Could you please clarify why did you implement this? I can't find where actually you need this.

**Review by vadim-zabolotniy — CHANGES_REQUESTED** [review]
Great work! Really. But have some room for improvements.

File: frontend/src/app/main/templates/TemplateModal/TemplateBuilder/ComponentsBuilder/ComponentSelectors/TypeSelector.js [hunk 32]
```diff
@@ -1,7 +1,7 @@
 import { FormControl, InputLabel, MenuItem, Select } from '@mui/material';
 import { useEffect, useState } from 'react';
 
-const TYPE_VALUES = ['helm_chart'];
+const TYPE_VALUES = ['helm_chart','curl'];
```
Comment (vadim-zabolotniy) [inline frontend/src/app/main/templates/TemplateModal/TemplateBuilder/ComponentsBuilder/ComponentSelectors/TypeSelector.js:4]: I'm not sure 100% but maybe it is some typo?

File: application/requirements.txt (hunk 21 above)
Comment (rtpro) [reply application/requirements.txt:15]: That's right, I'll pull it after this is merged.

File: application/requirements.txt (hunk 23 above)
Comment (rtpro) [reply application/requirements.txt:12]: Line 10

File: application/constants/templates.py (hunk 9 above)
Comment (rtpro) [reply application/constants/templates.py:81]: This is to be able to support both YAML list and a map, ie - the user can do
```
parameters:
- key1: value1
- key2: values2
```
or
```
parameters:
  key1: value1
  key2: values2
```

File: application/managers/http/manager.py (hunk 19 above)
Comment (rtpro) [reply application/managers/http/manager.py:51]: I renamed httpMethods to ComponentTypeHttpMethods to better represen that these types are for componentHttp.

File: .vscode/settings.json [hunk 33]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (vadim-zabolotniy) [inline .vscode/settings.json:1]: This setting shouldn't be in repository. Each developer should take care that his IDE stuff will not appear in repository.
All what you need to do is:
In `~/.gitconfig` set:
```
[core]
        excludesFile = ~/.gitignore
```
and in `~/.gitignore` add
```
.vscode
```

File: application/constants/http.py [hunk 34]
```diff
@@ -0,0 +1,37 @@
+"""
+Helm constants.
+"""
+from .base_enum import StrEnum
+
+
+class HttpStatuses(StrEnum):
+    """
+    Helm release statuses.
```
Comment (vadim-zabolotniy) [inline application/constants/http.py:9]: Helm? Did you revised statuses? How HTTP component can have status `pending-install` or `uninstalled`?

File: application/constants/http.py [hunk 35]
```diff
@@ -0,0 +1,37 @@
...
+    failed = 'failed'
+    uninstalling = 'uninstalling'
+    pending_install = 'pending-install'
+    pending_upgrade = 'pending-upgrade'
+    pending_rollback = 'pending-rollback'
+
+
+class HttpHealthStatuses(StrEnum):
+    """
+    Helm release operability statuses.
```
Comment (vadim-zabolotniy) [inline application/constants/http.py:34]: Helm?

File: application/constants/templates.py [hunk 36]
```diff
@@ -43,3 +53,62 @@ class HookTypes(StrEnum):
...
+    Template HTTP component methods.
+    """
+    get = 'get'
+    post = 'post'
+    put = 'put'
+    patch = 'patch'
+    delete = 'delete'
+
+
+class HttpDeploy(BaseModel):
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:69]: I just noticed that you have added Pydantic HTTP models to constants file. This is not proper place to define it. On my opinion it should be placed in `application/schemas/templates/components.py`.

File: application/constants/templates.py [hunk 37]
```diff
@@ -43,3 +53,62 @@ class HookTypes(StrEnum):
...
+    parameters: dict | list[dict] | None = Field(
+        description='Parameters for HTTP status', example=[{'param1': 'value1'}])
+
+
+class HttpDelete(BaseModel):
+    """
+    Delete schema for HTTP type components.
+    """
+    url: HttpUrl = Field(description='URL for HTTP delete.', example='http://example.com/path/to/delete')
+    method: ComponentTypeHttpMethods = Field(
```
Comment (vadim-zabolotniy) [inline application/constants/templates.py:106]: I'm not sure like 100% but please consider replacing `ComponentTypeHttpMethods` by `HTTPMethods` from `application/constants/common.py`.

File: application/constants/http.py (hunk 35 above)
Comment (vadim-zabolotniy) [reply application/constants/http.py:34]: I think adding of this health statuses it mistake. It will cause problem in application logic [here](https://github.com/JovianX/Service-Hub/blob/280-add-support-for-curl-component/application/managers/applications.py#L329) or [here](https://github.com/JovianX/Service-Hub/blob/280-add-support-for-curl-component/application/managers/applications.py#L345). In second example is my mistake I used `ApplicationHealthStatuses.healthy` instead `ReleaseHealthStatuses.unhealthy`.

What I'm suggesting. We should remove `HttpHealthStatuses` and rename `ReleaseHealthStatuses` to `ComponentHealthStatuses`. So it will reflect component status in general not Helm, HTTP or some other component type.

File: application/constants/templates.py (hunk 36 above)
Comment (vadim-zabolotniy) [reply application/constants/templates.py:69]: *After chat in telegram*

I think it should one Pydantic model but(parameter for all of them are identical) you need to add parameter to distinguish it for (un)install or status. Something like `trigger: ['install' | 'uninstall' | 'status']`.

File: application/requirements.txt (hunk 23 above)
Comment (vadim-zabolotniy) [reply application/requirements.txt:12]: Sorry. I forgot that we already used it. And was expecting you to add it in requirements.

File: application/schemas/templates/components.py [hunk 38]
```diff
@@ -28,8 +38,22 @@ class Component(BaseModel):
...
+            if deploy is None:
+                raise ValueError('deploy is required when type is http')
+            if delete is None:
+                raise ValueError('delete is required when type is http')
+        return values
+
     class Config:
         extra = Extra.forbid
 
-    def __hash__(self) -> str:
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:34]: I don't remember why I have added it. I think better to keep it.

File: application/schemas/templates/components.py [hunk 39]
```diff
@@ -28,8 +38,22 @@ class Component(BaseModel):
                     'a higher priority.',
     )
 
+    deploy: HttpDeploy | None = Field(description='Deployment details for HTTP type components')
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:41]: Need to separate helm and http components definitions. But to keep validation working we need to use `Field.discriminator` of Pydantic mode field. See how I did it for [Kubernetes entities](https://github.com/JovianX/Service-Hub/blob/main/application/services/kubernetes/schemas.py). Main trick does `Annotated` and `Field.discriminator`. So in `application/schemas/templates/template.py` you will have something like `components: Annotated[HelmComponent | HttpComponent, Field(discriminator='type')] | None` instead of `components: list[Component] | None`.

**Review by vadim-zabolotniy — CHANGES_REQUESTED** [review]
2nd review done.

File: application/constants/templates.py (hunk 37 above)
Comment (rtpro) [reply application/constants/templates.py:106]: I would like to have controll over the supported methods of ComponentTypeHttpMethods. ie  not all methods in HTTPMethods should be in HTTPMethods.

File: application/constants/http.py (hunk 35 above)
Comment (rtpro) [reply application/constants/http.py:34]: I think this should be handled by a seperate issue. Created issues https://github.com/JovianX/Service-Hub/issues/285

File: README.md [hunk 40]
```diff
@@ -120,7 +120,7 @@ For product documentation refer to the [Documentation Folder](documentation/READ
 
 # 📜 License
 
-Copyright 2022 JovianX Ltd.
+Copyright 2024 JovianX Ltd.
```
Comment (vadim-zabolotniy) [inline README.md:123]: Better to do such things in separate PR. ))

File: application/managers/http/schemas.py [hunk 41]
```diff
@@ -0,0 +1,3 @@
+from pydantic import BaseModel
```
Comment (vadim-zabolotniy) [inline application/managers/http/schemas.py:1]: This file is empty you can remove it. As I remember I had plans to keep all Pydantinc schemes  in `application/schemas`.

File: application/schemas/templates/components.py [hunk 42]
```diff
@@ -28,8 +41,89 @@ class Component(BaseModel):
                     'a higher priority.',
     )
 
-    class Config:
-        extra = Extra.forbid
+
+class ComponentTypeHttpMethods(StrEnum):
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:45]: This enum should be moved to `application/constants/http.py`

File: application/schemas/templates/components.py [hunk 43]
```diff
@@ -28,8 +41,89 @@ class Component(BaseModel):
                     'a higher priority.',
     )
 
-    class Config:
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:31]: You have lost setting that forbids additional parameters in Helm and HTTP components.

File: application/schemas/templates/components.py [hunk 44]
```diff
@@ -28,8 +41,89 @@ class Component(BaseModel):
...
+    delete: HttpRequest = Field(description='Deletion details for HTTP type components')
+    health: HttpRequest | None = Field(description='Health check for HTTP type components')
+
+
+Components = Annotated[ComponentHelm |
+                       ComponentHttp,
+                       Field(discriminator='type')]
+
+
+class Component(BaseModel):
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:89]: I think this model is redundant. You can do just `components: list[Components]` instead of `components: list[Component]` in `TemplateSchema` and validation should work fine. But don't trust me check it.))

Also you after removing `class Component(BaseModel):` you can rename `Components` to `Component`.

You can play with model and validation in real time:

1. `make setup`
2. `cd application/`
3. `python`
4. `from schemas.templates.template import TemplateSchema`
5. `with open('./path/to/template.yaml', 'r') as file:`
6. `    template = TemplateSchema.parse_obj(yaml.safe_load(file))` - pay attention on indent.

File: application/managers/http/manager.py [hunk 45]
```diff
@@ -0,0 +1,75 @@
...
+        """
+        # Convert to a dict if a list of dicts provided
+        if isinstance(parameters, list):
+            parameters = {k: v for d in parameters for k, v in d.items()}
+
+        # Create an HTTP client
+        async with httpx.AsyncClient() as client:
+            # Prepare the request
+            method = method.lower()
+            match method:
```
Comment (vadim-zabolotniy) [inline application/managers/http/manager.py:50]: Please no magic string. Refer to items from `ComponentTypeHttpMethods` enum. Or better use `httpx.request` get rid of this match-case statement. At this point here should be only HTTP method allowed in component model.

Comment (vadim-zabolotniy) [reply application/managers/http/manager.py:50]: [Docs](https://www.python-httpx.org/api/) about httpx.request.

File: application/schemas/templates/components.py (hunk 31 above)
Comment (rtpro) [reply application/schemas/templates/components.py:110]: This was needed for the install operation.

File: application/managers/http/manager.py (hunk 45 above)
Comment (rtpro) [reply application/managers/http/manager.py:50]: This allows the user to use UPPERCASE or lowercase in the tempalte.

File: application/schemas/templates/components.py [hunk 46]
```diff
@@ -28,8 +37,41 @@ class Component(BaseModel):
                     'a higher priority.',
     )
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:39]: You forgot to add:
```
    class Config:
        extra = Extra.forbid
```
This setting forbids unknown parameters in mode. By default in is silently ignored.

File: application/schemas/templates/components.py [hunk 47]
```diff
@@ -28,8 +37,41 @@ class Component(BaseModel):
...
+    delete: HttpRequest = Field(description='Deletion details for HTTP type components')
+    health: HttpRequest | None = Field(description='Health check for HTTP type components')
+
     class Config:
         extra = Extra.forbid
 
-    def __hash__(self) -> str:
-        return hash(self.name)
+
+Component = Annotated[ComponentHelm |
```
Comment (vadim-zabolotniy) [inline application/schemas/templates/components.py:75]: I think `Component` better to rename to `Components` because it represent multiple component not one.

**Review by vadim-zabolotniy — APPROVED** [review]
image
